File: lib_common/src/main/java/com/ebook/common/mvvm/viewmodel/BaseRefreshViewModel.java
package com.ebook.common.mvvm.viewmodel;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.databinding.ObservableArrayList;
import androidx.databinding.ObservableField;

import com.ebook.common.binding.command.BindingAction;
import com.ebook.common.binding.command.BindingCommand;
import com.ebook.common.event.SingleLiveEvent;
import com.ebook.common.mvvm.model.BaseModel;


public abstract class BaseRefreshViewModel<T, M extends BaseModel> extends BaseViewModel<M> {
    public ObservableField<Boolean> orientation = new ObservableField<>();
    public ObservableField<Boolean> enableLoadMore = new ObservableField<>();
    public ObservableField<Boolean> enableRefresh = new ObservableField<>();
    public BindingCommand onRefreshCommand = new BindingCommand(new BindingAction() {
        @Override
        public void call() {
            refreshData();
        }
    });
    public BindingCommand onLoadMoreCommand = new BindingCommand(new BindingAction() {
        @Override
        public void call() {
            loadMore();
        }
    });
    public BindingCommand onAutoRefreshCommand = new BindingCommand(new BindingAction() {
        @Override
        public void call() {
            refreshData();
        }
    });
    protected ObservableArrayList<T> mList = new ObservableArrayList<>();
    protected BaseRefreshViewModel.UIChangeRefreshLiveData mUIChangeRefreshLiveData;

    public BaseRefreshViewModel(@NonNull Application application, M model) {
        super(application, model);
        enableLoadMore.set(enableLoadMore());
        enableRefresh.set(enableRefresh());
    }

    public boolean enableLoadMore() {
        return true;
    }

    public boolean enableRefresh() {
        return true;
    }

    public BaseRefreshViewModel.UIChangeRefreshLiveData getUCRefresh() {
        if (mUIChangeRefreshLiveData == null) {
            mUIChangeRefreshLiveData = new BaseRefreshViewModel.UIChangeRefreshLiveData();
        }
        return mUIChangeRefreshLiveData;
    }

    public void postStopRefreshEvent() {
        if (mUIChangeRefreshLiveData != null) {
            mUIChangeRefreshLiveData.getStopRefresLiveEvent().call();
        }
    }

    public void postAutoRefreshEvent() {
        if (mUIChangeRefreshLiveData != null) {
            mUIChangeRefreshLiveData.getAutoRefresLiveEvent().call();
        }
    }

    public void postStopLoadMoreEvent() {
        if (mUIChangeRefreshLiveData != null) {
            mUIChangeRefreshLiveData.mStopLoadMoreLiveEvent.call();
        }
    }

    public ObservableArrayList<T> getList() {
        return mList;
    }

    public abstract void refreshData();

    public abstract void loadMore();

    public final class UIChangeRefreshLiveData extends SingleLiveEvent {
        private SingleLiveEvent<Void> mStopRefresLiveEvent;
        private SingleLiveEvent<Void> mAutoRefresLiveEvent;
        private SingleLiveEvent<Void> mStopLoadMoreLiveEvent;

        public SingleLiveEvent<Void> getStopRefresLiveEvent() {
            return mStopRefresLiveEvent = createLiveData(mStopRefresLiveEvent);
        }

        public SingleLiveEvent<Void> getAutoRefresLiveEvent() {
            return mAutoRefresLiveEvent = createLiveData(mAutoRefresLiveEvent);
        }

        public SingleLiveEvent<Void> getStopLoadMoreLiveEvent() {
            return mStopLoadMoreLiveEvent = createLiveData(mStopLoadMoreLiveEvent);
        }
    }

}

File: lib_common/src/main/java/com/xrn1997/common/mvvm/viewmodel/BaseRefreshViewModel.kt
package com.xrn1997.common.mvvm.viewmodel

import android.app.Application
import android.util.Log
import androidx.databinding.ObservableArrayList
import androidx.lifecycle.MutableLiveData
import com.xrn1997.common.adapter.binding.BindingAction
import com.xrn1997.common.adapter.binding.BindingCommand
import com.xrn1997.common.event.SingleLiveEvent
import com.xrn1997.common.mvvm.model.BaseModel

/**
 * 可刷新的ViewModel基类
 * @author xrn1997
 */
@Suppress("unused")
abstract class BaseRefreshViewModel<T, M : BaseModel>(
    application: Application,
    model: M
) : BaseViewModel<M>(application, model) {
    private var list: ObservableArrayList<T> = ObservableArrayList()
    val mList: ObservableArrayList<T>
        get() = list
    var orientation: MutableLiveData<Boolean> = MutableLiveData()
    var enableLoadMore: MutableLiveData<Boolean> = MutableLiveData()
    var enableRefresh: MutableLiveData<Boolean> = MutableLiveData()
    protected var defaultUIChangeRefreshLiveData: UIChangeRefreshLiveData? = null
    val mUIChangeRefreshLiveData: UIChangeRefreshLiveData
        get() {
            if (defaultUIChangeRefreshLiveData == null) {
                defaultUIChangeRefreshLiveData = UIChangeRefreshLiveData()
            }
            return defaultUIChangeRefreshLiveData as UIChangeRefreshLiveData
        }

    inner class UIChangeRefreshLiveData {
        private var stopRefreshLiveEvent: SingleLiveEvent<Boolean>? = null
        private var autoRefreshLiveEvent: SingleLiveEvent<Void>? = null
        private var stopLoadMoreLiveEvent: SingleLiveEvent<Boolean>? = null
        val mStopRefreshLiveEvent
            get() = createLiveData(stopRefreshLiveEvent).also { stopRefreshLiveEvent = it }
        val mAutoRefreshLiveEvent
            get() = createLiveData(autoRefreshLiveEvent).also { autoRefreshLiveEvent = it }
        val mStopLoadMoreLiveEvent
            get() = createLiveData(stopLoadMoreLiveEvent).also { stopLoadMoreLiveEvent = it }
    }

    /**
     *
     * 停止刷新
     * @param boolean Boolean 数据是否成功刷新 （会影响到上次更新时间的改变）
     */
    fun postStopRefreshEvent(boolean: Boolean) {
        mUIChangeRefreshLiveData.mStopRefreshLiveEvent.postValue(boolean)
    }

    /**
     * 自动刷新
     */
    fun postAutoRefreshEvent() {
        mUIChangeRefreshLiveData.mAutoRefreshLiveEvent.call()
    }

    /**
     * 停止加载
     * @param boolean Boolean 数据是否成功
     */
    fun postStopLoadMoreEvent(boolean: Boolean) {
        mUIChangeRefreshLiveData.mStopLoadMoreLiveEvent.postValue(boolean)
    }


    /**
     * 刷新命令
     */
    var onRefreshCommand = BindingCommand(object : BindingAction {
        override fun call() {
            Log.e("BRViewModel", "下拉刷新")
            refreshData()
        }
    })

    /**
     * 加载命令
     */
    var onLoadMoreCommand = BindingCommand(object : BindingAction {
        override fun call() {
            Log.e("BRViewModel", "上拉加载")
            loadMore()
        }
    })

    /**
     * 刷新数据
     */
    abstract fun refreshData()

    /**
     * 加载更多
     */
    open fun loadMore() {}

    /**
     * 是否启用上拉加载
     * @return Boolean 默认false
     */
    open fun enableLoadMore(): Boolean {
        return false
    }

    /***
     * 是否启用下拉刷新
     * @return Boolean 默认true
     */
    open fun enableRefresh(): Boolean {
        return true
    }

    init {
        val loadMore by lazy { enableLoadMore() }
        enableLoadMore.value = loadMore
        val refresh by lazy { enableRefresh() }
        enableRefresh.value = refresh
    }
}